tripleo/insights/git.py

```diff
@@ -203,3 +208,81 @@ def _get_new_cli_downloader(self, url: URL) -> GitCLIDownloader:
 
     def _get_new_github_downloader(self, url: URL) -> GitHubDownloader:
         return GitHubDownloader(repository=url)
+
+
+class GitDownload:
+    """Takes care of joining all the tools in this module together and
+    allow to safely download files from a Git repository.
+    """
+
+    def __init__(
+        self,
+        downloader_fetcher: GitDownloaderFetcher = GitDownloaderFetcher()
+    ):
+        """Constructor.
+
+        :param downloader_fetcher: Tool that provides the APIs that this
+            will use to interact with Git.
+        """
+        self._download_fetcher = downloader_fetcher
+
+    @property
+    def download_fetcher(self) -> GitDownloaderFetcher:
+        """
+        :return: Tool that provides the APIs that this
+            will use to interact with Git.
+        """
+        return self._download_fetcher
+
+    def download_as_yaml(
+        self,
+        repo: URL,
+        file: str,
+        yaml_parser: YAMLParser = StandardYAMLParser()
+    ) -> YAML:
+        """Downloads a YAML file from a Git repository and parses it into
+        a Python object.
+
+        :param repo: The Git repository.
+        :param file: Relative path to the repo's root to the file to download.
+        :param yaml_parser: Tool used to parse the YAML file.
+        :return: The file as a Python object.
+        :raises DownloadError:
+            If no API is available to download the file.
+            If all APIs failed to download the file.
+            If the file is not in YAML format.
+        """
+        try:
+            for api in self._get_apis_for(repo):
+                LOG.info(
+                    "Trying to download file: '%s' with API: '%s'...",
+                    file, type(api).__name__
+                )
+
+                try:
+                    return yaml_parser.as_yaml(api.download_as_text(file))
+                except YAMLError as ex:
+                    msg = f"Failed to parse YAML file: '{file}'."
+                    raise DownloadError(msg) from ex
```

Comment:
If I understood correctly, the goal here is to try the different API's until one can provide the file, but raise an error if all fail. Won't raising this exception interrrupt this flow? If for example the first api had a malformed copy of the file but the second had a good copy, the second would never be queried, would it?

Reply:
I do not know how common that would be to happen. The repository and file are the same for all APIs, so they should all be talking about the same thing. There may be some corruption during the file's download, but is that something that we would like to consider? 

The idea behind this check is that if an API sees the file is not YAML, do not bother trying the others, they will all fail.

Reply:
I agree that it would be a really infrequent corner case, and I'm not really worried about it, just wanted to make sure that I understood what the function intended to  do. As you say, it's perfectly reasonable to stop if the file is not YAML.

Reply:
Oh, I see. I read the comment wrong before.